Add env var to override OCCA HIP compile arch

dmcdougall opened the pull request:
## Description

This commit teaches OCCA to understand the `OCCA_HIP_OVERRIDE_COMPILE_ARCH` environment variable.  If set, OCCA will ignore the arch set in the kernel properties and try to compile for the architecture provided in the environment variable.

This allows the user to have more control over what targets are JITted for. Usecases are for JITting for generic targets (see [here](https://github.com/llvm/llvm-project/commit/f93aa5157a3317b24cff660ac972814ee9ed4dbc)) and for the (currently beta) amdgcnspirv target (see [here](https://rocm.docs.amd.com/projects/llvm-project/en/latest/conceptual/spirv.html#rocm-support-for-spir-v-beta)).

Unfortunately, the amdgcnspirv target name doesn't start with "gfx", so I added a branch that checks for this case and treats it accordingly.

Comments, criticism, tomatoes, and hecklers are not only welcome but heavily encouraged.

<!-- Thank you for contributing! -->

dmcdougall committed:
Add env var to override OCCA HIP compile arch

This commit teaches OCCA to understand the OCCA_HIP_OVERRIDE_COMPILE_ARCH
environment variable.  If set, OCCA will ignore the arch set in the kernel
properties and try to compile for the architecture provided in the environment
variable.

This allows the user to have more control over what targets are JITted for.
Usecases are for JITting for generic targets (see here:
https://github.com/llvm/llvm-project/commit/f93aa5157a3317b24cff660ac972814ee9ed4dbc)
and for the (currently beta) amdgcnspirv target (see here:
https://rocm.docs.amd.com/projects/llvm-project/en/latest/conceptual/spirv.html#rocm-support-for-spir-v-beta).

Unfortunately, the amdgcnspirv target name doesn't start with "gfx", so I added
a branch that checks for this case and treats it accordingly.

thilinarmtb commented:
@dmcdougall : Seems like both CUDA and HIP both set the architecture compiler option. Wondering
if it is a good idea to rename the variable to `OCCA_OVERRIDE_COMPILE_ARCH` and let the backends which
set this compiler option query the env var (or better yet we update the kernelProperties based on the
env var and pass it around).

kris-rowe commented:
@dmcdougall Could you update this to target the development branch?

dmcdougall commented:
> Could you update this to target the development branch?

Rebased on top of `development` and now targets `development`.

dmcdougall commented:
> Seems like both CUDA and HIP both set the architecture compiler option.

Honestly, I think any backend should be able to do this.

> Wondering if it is a good idea to rename the variable to OCCA_OVERRIDE_COMPILE_ARCH and let the backends which
set this compiler option query the env var

I totally agree.  I had initially wanted to do it this way, but didn't see a way for occa to consume the variable in a place that wasn't backend-specific.

> (or better yet we update the kernelProperties based on the env var and pass it around).

Ah yes, maybe this is they way I should have done it to begin with.

dmcdougall commented:
Before I continue with implementing @thilinarmtb's feedback, is the general principle behind this idea acceptable?

thilinarmtb commented:
> Before I continue with implementing @thilinarmtb's feedback, is the general principle behind this idea acceptable?

I think this is a useful environment variable to have. Kris is on vacation so probably we will have to
wait till next week to know his opinion.

dmcdougall committed:
Make OCCA_HIP_OVERRIDE_COMPILE_ARCH a generic name

Now each backend can consume it and the user understands the override
only applies to the selected backend at runtime.

dmcdougall commented:
I decided to just rename it instead, and let the backend be responsible for deciding how to populate the kernel properties.
